executor/runcexecutor/executor.go

```diff
@@ -310,6 +315,62 @@ func (w *runcExecutor) Exec(ctx context.Context, meta executor.Meta, root cache.
 	return nil
 }
 
+func (w *runcExecutor) Exec(ctx context.Context, id string, process executor.ProcessInfo) error {
+	// first verify the container is running, if we get an error assume the container
```

Comment:
Could we do some sync with go mutex/cond/channel in here first. Eg. detect that `Run()` has been called for the ID and it has not returned yet. I guess there is still a race when `runc` binary is being called that needs to be checked with retries but first check and reacting to pid1 returning would be fast then. Also, maybe`Run()` should take a channel that is closed to indicate started event? Ideally, when pid1 returns while this method is running all the errors returned in here would be about pid1 closing, and not for example that open fails because state file has been deleted.

Reply:
yeah, I struggled a bit with race conditions between pid1 and exec.  I think a channel passed to Run seems reasonable.  Since runc.Runc is blocking I can setup a short goroutine to poll for "running" state before sending on the channel to avoid the ~200ms delay between runc.Run being called and the container being running.  I will try to rework it a bit to avoid the Exec retry loop.

Reply:
Something like

```
Run(id string, started chan<- struct{}) {
  done = make(chan error, 1)
  
  e.mu.Lock()
  e.m[id] = done
  defer {
    e.mu.Lock()
    delete(e.m, id)
    e.mu.Unlock()
    done <- err
  }()
  e.mu.Unlock()
  
  close(started)
}

Exec(id string) {
  e.mu.Lock()
  done, ok := e.m[id]
  if !ok {
    return error404
  }
  e.mu.Unlock()
  
  defer func() {
    select {
    case err := <-done:
      err = errors.Wrapf(err, "pid1 returned with error")
    default:
    }
  }()
}
``` 
?

I still think retry is needed as well because of the way we call runc without getting the create event.

Reply:
makes sense, I will update.